### ISSUE TITLE
Gaps and overlaps in tiles on high zoom with -zg

### ISSUE THREAD

**Issue by jdeboer-geoplan** [mapbox/tippecanoe#562]
Hi,  I've got a small problem with the -zg option, there might be a flag option I'm missing that would sort it out.  When I run with -zg it guesses that I want to go down to z8

Choosing a maxzoom of -z6 for features about 5147 feet (1569 meters) apart
Choosing a maxzoom of -z8 for resolution of about 1281 feet (390 meters) within features

However when I zoom in to level 14 it has lots of holes and overlaps, if I generate it to level 14 it looks fine all the way to 20.

Output
![image](https://user-images.githubusercontent.com/23341752/39467794-c58917a4-4d27-11e8-8f5a-ef9e73314e80.png)

If I use -ab the results are much better but still not perfect:
Output
![image](https://user-images.githubusercontent.com/23341752/39467906-5fb40da2-4d28-11e8-8877-c76d3af47236.png)

I'd hoped -pS might help but although the tile file is slightly bigger I can't see any difference in output.

I can provide the geoJSON but I feel this must be a known problem and I'm just failing to find the right flag combo to sort it out.  Obviously I can get my tiles by generating all the way to z14 but I'm attracted to the much faster creation and smaller tile file size of only going to z8 if that's possible with my data.

Thanks

**Comment by flippmoke** [comment]
This is due to simplification more then likely. If two polygons share an edge and then are simplified in a very slightly different way there is a good change you get output like what you see above. You might try playing with these settings: https://github.com/mapbox/tippecanoe#line-and-polygon-simplification

**Comment by jdeboer-geoplan** [comment]
thanks @flippmoke I tried -ps and pS but it didn't help, still get the gaps.  What value could I use in the --simplification= to have a better scale?

**Comment by pathmapper** [comment]
@jdeboer-geoplan did you try the following? 

`-ab` or  `--detect-shared-borders`

From the [docs](https://github.com/mapbox/tippecanoe#attempts-to-improve-shared-polygon-boundaries):

"... detect borders that are shared between multiple polygons and **simplify them identically in each polygon.**"

**Comment by jdeboer-geoplan** [comment]
HI @pathmapper yup, that dramatically improved things as you can see in the difference between pic1 and pic2.

**Comment by pathmapper** [comment]
@jdeboer-geoplan ahh, sorry, I missed that.

**Comment by jdeboer-geoplan** [comment]
Absolutely.  I'll attache a really cut down copy of my geoJson.  One thing I've noticed is that -zg goes higher with less data. e.g
Whole uk - Choosing a maxzoom of -z8 for resolution of about 1281 feet (390 meters) within features
Small part of london - Choosing a maxzoom of -z9 for features about 509 feet (156 meters) apart
two streets - Choosing a maxzoom of -z10 for resolution of about 475 feet (145 meters) within features

**Comment by jdeboer-geoplan** [comment]
[1998_UK_GEO_GSSP_cutdown.zip](https://github.com/mapbox/tippecanoe/files/2009993/1998_UK_GEO_GSSP_cutdown.zip)

**Comment by ericfischer** [comment]
`-zg` works from whatever the geographic density is in the input data, so if your smaller sample has denser or more detailed geometries than the full data, it will choose a higher zoom level based on that. (London in particular may also throw off the calculation, since the Prime Meridian cuts through it and complicates the tile math.)

Thanks for the data. I'll take a look now.

**Comment by jdeboer-geoplan** [comment]
Hi @ericfischer thanks for taking a look, on the -zg thing the geometry was the same, just I'd removed a bunch of them, is that worth raising as a separate issue to these holes from polygon cleanup?

**Comment by ericfischer** [comment]
`-zg` is based on the *average* precision of the features, not the *highest* precision, so it is not necessarily surprising that it changes when you change the sample. But if you think it is doing a bad job of choosing, please do file another ticket with a data file and what you think it should have chosen for that data.

**Comment by ericfischer** [comment]
Alternately, an `--extend-zooms-if-still-simplifying` would probably be more practical than trying to predict up front what will get snap-rounded.

**Comment by ericfischer** [comment]
Even just checking for snap rounding that happens within the tile proper (as opposed to in the buffer), continuing until there is no snapping keeps tiling far too deeply in cases like the Natural Earth countries, which are still doing some snapping at z11.

**Comment by jdeboer-geoplan** [comment]
wow, this is a can of worms.  What snap rounding algo is wagyu using? I assume it's not idempotent?  I like the idea of --extend-zooms-if-still-simplifying

**Comment by jdeboer-geoplan** [comment]
Hi, well I'm learning loads reading up about the exciting array of different snap rounding algos out there, good stuff but not necessarily of any use here.  I think that the global snap rounding pass sounds good, that will presumably mean that there maybe more nodes on each polygon and so they might be a little bigger?

**Comment by ericfischer** [comment]
Yes, they'll get bigger because there will be more nodes.

I'm actually much more worried about time and memory than about tile size, especially in the future (https://github.com/mapbox/vector-tile-spec/issues/104) when features can be rejoined across tile boundaries, in which case a large polygon will need to be snapped against every feature that it touches in every tile, not just in one.

**Comment by Plantain** [comment]
I think I'm seeing the same behaviour, and I have a JSON file that reproduces it quite nicely.
<img width="907" alt="screen shot 2018-06-05 at 10 01 43 am" src="https://user-images.githubusercontent.com/319895/40963038-a22373cc-68a7-11e8-8314-c458bb20a8e2.png">
<img width="891" alt="screen shot 2018-06-05 at 10 01 31 am" src="https://user-images.githubusercontent.com/319895/40963039-a25d045c-68a7-11e8-9f10-48493f45d7ee.png">
[1700.zip](https://github.com/mapbox/tippecanoe/files/2071380/1700.zip)

tippecanoe is multiple orders of magnitude faster than our previous tile generator, so time and memory isn't a problem for us yet! :) :) :)

**Comment by Plantain** [comment]
>"It does look like the gaps are being introduced during polygon cleanup with Wagyu. I don't see them with cleanup turned off."
@ericfischer, is the Wagyu polygon cleanup controlled by a flag? Otherwise can you share how you disabled it so we can work around this in the meantime?

**Comment by ericfischer** [comment]
It is not controlled by a flag because the vector tile spec requires that polygon geometries be valid, and I don't have a way to guarantee that without the use of Wagyu. But if you want to live dangerously, you can comment out the

```
geoms[g] = clean_or_clip_poly(geoms[g], 0, 0, false);
```

line here: https://github.com/mapbox/tippecanoe/blob/d64ac19f115bef509f34cfc685ff27644e033e3c/tile.cpp#L496

**Comment by ericfischer** [comment]
And please note that `--coalesce-smallest-as-needed` will only be useful *with* the use of Wagyu, since its entire value relies on being able to union together adjacent polygons, which happens in the course of the clean-or-clip call to Wagyu.

**Comment by Plantain** [comment]
For comparison, mapshaper shows a lot of the whitespace is 'real', but the weird overlapped artifacts aren't there.
<img width="846" alt="screen shot 2018-06-20 at 8 04 52 pm" src="https://user-images.githubusercontent.com/319895/41676274-44a77bc4-74c5-11e8-8ec1-d3b9b92b365a.png">

**Comment by ericfischer** [comment]
Interesting. Can you share a sample of your data so I can investigate?

**Comment by Plantain** [comment]
Attached above - 1800.gz.

**Comment by Plantain** [comment]
The zoom in particular is from this location:
<img width="846" alt="screen shot 2018-06-20 at 9 04 35 pm" src="https://user-images.githubusercontent.com/319895/41679211-b9fd9e82-74cd-11e8-8273-fb85dbca5c3f.png">

**Comment by ericfischer** [comment]
For my reference: the area in question is near 9.5/43.9681/-112.0949.

I'm pretty sure the geometric errors are in your original data, which is only given to 3 decimal digits of precision (about 350 feet). If I convert your polygon rings to individual LineStrings and tile them considerably deeper than the display resolution, it is clear that the contours do not match up in many places.

![screen shot 2018-06-20 at 2 45 16 pm](https://user-images.githubusercontent.com/1951835/41686394-c0d8b870-7498-11e8-8c1a-4aa9c9113357.png)

If you are not seeing overlaps in mapshaper, it is because it is drawing the shapes opaquely rather than with partial transparency, which hides the overlaps.

I think whatever process is creating these polygons is doing its own line simplification, and that to get the results you want you will need to stop it from simplifying or get it to use a higher precision during simplification.

**Comment by Plantain** [comment]
You were right - there was some line simplification going on in the geojson library generating it.

**Comment by stonetip** [comment]
In many cases, simplifying adjacent features leads to overlaps and gaps if the topology isn't good in the first place, i.e. adjacent features don't share vertices. This needs to be corrected in your source data before exporting to GeoJSON and tiles. For example, you can use tools like v.clean in QGIS to try to automatically clean up topology.
![image](https://user-images.githubusercontent.com/3766798/72690593-fd967900-3ada-11ea-911f-831f6d705e91.png)

**Comment by briantran37** [comment]
im getting a similar issue but my vector tile is pixelated at high zoom levels. 


~/bin/tippecanoe -r1 -Z9 -d18 -o output.mbtiles --force --layer=layername --detect-shared-borders --extend-zooms-if-still-dropping --drop-fraction-as-needed --drop-densest-as-needed input.geojson

![image](https://user-images.githubusercontent.com/72456221/101271864-d37e9180-3754-11eb-94b5-f8afd25afd44.png)

how do i make it smoother?
